README.md
# CounterspellAlert
A simple add-on that announces when you are afflicted by a counterspell/crowd control effect.

![Preview of addon in action.](https://silverhawke.s-ul.eu/7l3NAVvw)

## Installing
Downloading from [CurseForge](https://www.curseforge.com/wow/addons/counterspellalert) is recommended.
Alternatively, download or clone from GitHub.

## Configuration
Type `/csalert` or go to `Main Menu → Interface Options → AddOns → CounterspellAlert`.

## License
Copyright © 2020 silverhawke | [MIT License](https://opensource.org/licenses/MIT)


changelog.txt
v1.0.2:
- Fixed error when addon is run for the first time.

v1.0.1:
- Better handling of saved settings.
- Restoring to default settings no longer cause an error.

v1.0.0:
- Initial release.


Helpers.lua
local _, CsAlert = ...

CsAlert.func = {}

-- Announces messages
function CsAlert.func.announce(message, target)
    if CsAlert_GlobalSettings.testMode then
        print("|cffff7f3f<CsAlert>|r " .. message)
    else
        SendChatMessage(message, target)
    end
end

-- Converts flag values to chat-friendly raid target marker
function CsAlert.func.getRaidTargetString(flag)
    if flag == 0x01 then
        return "rt1"
    elseif flag == 0x02 then
        return "rt2"
    elseif flag == 0x04 then
        return "rt3"
    elseif flag == 0x08 then
        return "rt4"
    elseif flag == 0x10 then
        return "rt5"
    elseif flag == 0x20 then
        return "rt6"
    elseif flag == 0x40 then
        return "rt7"
    elseif flag == 0x80 then
        return "rt8"
    else
        return ""
    end
end

-- Iterates over a table sorted by keys
function CsAlert.func.sortedKeyIter(t, comp)
    local keys = {}
    for k, _ in pairs(t) do
        tinsert(keys, k)
    end
    sort(keys, comp)

    local i = 0
    local iter = function()
        i = i + 1
        if keys[i] == nil then return nil
        else return keys[i], t[keys[i]]
        end
    end

    return iter
end

-- Creates a header with a line under it
function CsAlert.func.createHeader(text, parent, ofsx, ofsy, fontObj)
    local fs = parent:CreateFontString(nil, "ARTWORK")
    fontObj = fontObj or GameFontNormalLarge
    fs:SetFontObject(fontObj)
    fs:SetText(text)
    fs:SetPoint("TOPLEFT", parent, "TOPLEFT", ofsx, ofsy)

    local line = parent:CreateLine()
    line:SetColorTexture(1, 1, 1, 0.3)
    line:SetStartPoint("TOPLEFT", fs, 0, -3 - fs:GetStringHeight())
    line:SetEndPoint("TOPRIGHT", fs, 5, -3 - fs:GetStringHeight())
    line:SetThickness(1.5)

    return fs
end

-- Sorts chatTypeIds in a particular order
local chatTypeIdOrder = {NONE=1, SAY=2, YELL=3, PARTY=4, RAID=5, INSTANCE_CHAT=6}
function CsAlert.func.sortChatTypeId(a, b)
    return chatTypeIdOrder[a] < chatTypeIdOrder[b]
end

-- Copies tables, making sure to copy embedded tables too
-- Lifted from lua-users.org
function CsAlert.func.deepCopy(orig)
    local orig_type = type(orig)
    local copy
    if orig_type == 'table' then
        copy = {}
        for orig_key, orig_value in next, orig, nil do
            copy[CsAlert.func.deepCopy(orig_key)] = CsAlert.func.deepCopy(orig_value)
        end
        setmetatable(copy, CsAlert.func.deepCopy(getmetatable(orig)))
    else -- number, string, boolean, etc
        copy = orig
    end
    return copy
end

-- Copies values from t2 to t1, recursively
-- With an option to only overwrite nils
function CsAlert.func.updateTable(t1, t2, onlyNilValues)
    for k, v in pairs(t2) do
        if type(v) == "table" then
            if t1[k] == nil then t1[k] = {} end
            CsAlert.func.updateTable(t1[k], v)
        else
            if not onlyNilValues or t1[k] == nil then
                t1[k] = v
            end
        end
    end
end

-- Drops entries in t1 if the key doesn't exist in t2
-- Returns a copy -- doesn't modify in-place
-- Also filters recursively!
function CsAlert.func.filterTable(t1, t2)
    local t = {}
    for k, v in pairs(t1) do
        if t2[k] ~= nil then
            if type(v) == "table" then
                t[k] = CsAlert.func.filterTable(v, t2[k])
            else
                t[k] = v
            end
        end
    end
    return t
end

-- Prints a table to default chat frame
function CsAlert.func.dumpTable(t, l)
    local indent = ""
    local initBrace = l == nil
    l = l or 1
    for _ = 1, l do
        indent = indent .. "  "
    end
    if initBrace then print(indent .. "{") end
    for k, v in CsAlert.func.sortedKeyIter(t) do
        if type(v) ~= "table" then
            print(indent .. "  ", k .. ":", v)
        else
            print(indent .. "  ", k .. ": {")
            CsAlert.func.dumpTable(v, l + 1)
        end
    end
    print(indent .. "}")
end

-- Updates options frame settings
function CsAlert.func.updateOptions(globalSet, charaSet)
    -- Global settings
    CsAlert.optionsFrame.checkTestMode:SetChecked(globalSet.testMode)
    CsAlert.optionsFrame.checkShowAllDebuff:SetChecked(globalSet.showAllDebuff)
    if not globalSet.testMode then
        CsAlert.optionsFrame.checkShowAllDebuff:Disable()
        CsAlert.optionsFrame.checkShowAllDebuff.text:SetFontObject(GameFontDisableSmall)
    end

    -- Per character settings, setting defaults if needed
    CsAlert.optionsFrame.editMinDebuffDuration:SetNumber(charaSet.minDebuffDuration)
    CsAlert.optionsFrame.editMinDebuffDuration:SetCursorPosition(0)
    charaSet.shownDebuffType = charaSet.shownDebuffType or CsAlert.charaSettings.shownDebuffType
    for key, checkbox in pairs(CsAlert.optionsFrame.checkDebuffTypes) do
		checkbox:SetChecked(charaSet.shownDebuffType[key])
    end
    charaSet.broadcastSettings = charaSet.broadcastSettings or CsAlert.charaSettings.broadcastSettings
    for key, dropdown in pairs(CsAlert.optionsFrame.dropDownBroadcast) do
		local v = charaSet.broadcastSettings[key]
        UIDropDownMenu_SetText(dropdown, CsAlert.strings.broadcastTargets[v])
    end
end

Options.lua
local _, CsAlert = ...

-- Create options frame
local optionsFrame = CreateFrame("Frame", nil, UIParent)
optionsFrame.name = "CounterspellAlert"
optionsFrame.settingChanges = {globalSettings={}, charaSettings={}}
function optionsFrame.okay()
    CsAlert.func.updateTable(CsAlert_GlobalSettings, optionsFrame.settingChanges.globalSettings)
    CsAlert.func.updateTable(CsAlert_CharaSettings, optionsFrame.settingChanges.charaSettings)
    for _, t in pairs(optionsFrame.settingChanges) do
        wipe(t)
    end
end
function optionsFrame.cancel()
    for _, t in pairs(optionsFrame.settingChanges) do
        wipe(t)
    end
    CsAlert.func.updateOptions(CsAlert_GlobalSettings, CsAlert_CharaSettings)
end
function optionsFrame.default()
    optionsFrame.settingChanges = {
        globalSettings = CsAlert.func.deepCopy(CsAlert.globalSettings),
        charaSettings = CsAlert.func.deepCopy(CsAlert.charaSettings)
    }
    CsAlert.func.updateOptions(optionsFrame.settingChanges.globalSettings, optionsFrame.settingChanges.charaSettings)
end
CsAlert.optionsFrame = optionsFrame
InterfaceOptions_AddCategory(optionsFrame)

-- Global settings header
local text = CsAlert.func.createHeader(CsAlert.strings.globalSettings, optionsFrame, 20, -20)

-- Test mode check button
local checkButton = CreateFrame("CheckButton", nil, optionsFrame, "UICheckButtonTemplate")
checkButton.text:SetText(CsAlert.strings.testMode)
checkButton:SetPoint("TOPLEFT", text, "BOTTOMLEFT", 0, -20)
checkButton:SetScript("OnClick", function(self)
    local changes = optionsFrame.settingChanges.globalSettings
    changes.testMode = self:GetChecked()
    if changes.testMode then
        optionsFrame.checkShowAllDebuff:Enable()
        optionsFrame.checkShowAllDebuff.text:SetFontObject(GameFontNormalSmall)
    else
        changes.showAllDebuff = false
        optionsFrame.checkShowAllDebuff:SetChecked(false)
        optionsFrame.checkShowAllDebuff:Disable()
        optionsFrame.checkShowAllDebuff.text:SetFontObject(GameFontDisableSmall)
    end
end)
checkButton:SetScript("OnEnter", function(self)
    GameTooltip:SetOwner(self, "ANCHOR_BOTTOMRIGHT")
    GameTooltip:AddLine(CsAlert.strings.testMode)
    GameTooltip:AddLine(CsAlert.strings.testModeDesc, 1, 1, 1, true)
    GameTooltip:Show()
end)
checkButton:SetScript("OnLeave", function()
    GameTooltip:Hide()
end)
optionsFrame.checkTestMode = checkButton
local checkTestModeLength = optionsFrame.checkTestMode.text:GetStringWidth()

-- Show all debuffs check button
checkButton = CreateFrame("CheckButton", nil, optionsFrame, "UICheckButtonTemplate")
checkButton.text:SetText(CsAlert.strings.showAllDebuff)
checkButton:SetPoint("TOPLEFT", optionsFrame.checkTestMode, "TOPRIGHT", 10 + checkTestModeLength, 0)
checkButton:SetScript("OnClick", function(self)
    local changes = optionsFrame.settingChanges.globalSettings
    changes.showAllDebuff = self:GetChecked()
end)
checkButton:SetScript("OnEnter", function(self)
    GameTooltip:SetOwner(self, "ANCHOR_BOTTOMRIGHT")
    GameTooltip:AddLine(CsAlert.strings.showAllDebuff)
    GameTooltip:AddLine(CsAlert.strings.showAllDebuffDesc, 1, 1, 1, true)
    GameTooltip:Show()
end)
checkButton:SetScript("OnLeave", function()
    GameTooltip:Hide()
end)
optionsFrame.checkShowAllDebuff = checkButton

-- Character-specific settings header
local prevText = CsAlert.func.createHeader(CsAlert.strings.charaSettings, optionsFrame.checkTestMode, 0, -60)

-- Min debuff duration text
text = optionsFrame:CreateFontString(nil, "ARTWORK")
text:SetFontObject(GameFontNormalSmall)
text:SetText(CsAlert.strings.minDebuffDuration .. ":")
text:SetPoint("TOPLEFT", prevText, "BOTTOMLEFT", 0, -32)
local textMinDebuffDurationLength = text:GetStringWidth()

-- Min debuff duration edit box
local editBox = CreateFrame("EditBox", nil, optionsFrame)
editBox:SetBackdrop({
    bgFile = "Interface/Tooltips/UI-Tooltip-Background",
    edgeFile = "Interface/Tooltips/UI-Tooltip-Border",
    tile = true,
    tileSize = 26,
    edgeSize = 16,
    insets = { left = 4, right = 4, top = 4, bottom = 4}
})
editBox:SetBackdropColor(0,0,0,1)
editBox:SetPoint("TOPLEFT", prevText, "BOTTOMLEFT", 10 + textMinDebuffDurationLength, -25)
editBox:SetSize(40, 25)
editBox:SetMultiLine(false)
editBox:SetAutoFocus(false)
editBox:SetJustifyH("CENTER")
editBox:SetJustifyV("CENTER")
editBox:SetFontObject(GameFontWhiteSmall)
editBox:SetNumeric(true)
editBox:SetScript("OnTextChanged", function(self)
    local changes = optionsFrame.settingChanges.charaSettings
    changes.minDebuffDuration = self:GetNumber()
end)
editBox:SetScript("OnEnterPressed", function(self)
    self:ClearFocus()
end)
editBox:SetScript("OnEscapePressed", function(self)
    self:ClearFocus()
end)

-- Min debuff duration edit box's extra text
editBox.title_text = editBox:CreateFontString(nil, "ARTWORK")
editBox.title_text:SetFontObject(GameFontNormalSmall)
editBox.title_text:SetText(CsAlert.strings.seconds)
editBox.title_text:SetPoint("TOPLEFT", editBox, "TOPRIGHT", 5, -7)
optionsFrame.editMinDebuffDuration = editBox

-- Types of debuff to announce
text = optionsFrame:CreateFontString(nil, "ARTWORK")
text:SetFontObject(GameFontNormalSmall)
text:SetText(CsAlert.strings.debuffTypeText .. ":")
text:SetPoint("TOPLEFT", editBox, "BOTTOMLEFT", -(10 + textMinDebuffDurationLength), -20)

optionsFrame.checkDebuffTypes = {}
local counter, firstKey = 0, nil
for k, s in CsAlert.func.sortedKeyIter(CsAlert.strings.statuses) do
    if firstKey == nil then firstKey = k end
    local checkBox = CreateFrame("CheckButton", nil, optionsFrame, "UICheckButtonTemplate")
    local x, y = counter % 3, floor(counter / 3)
    checkBox:SetPoint("TOPLEFT", text, "BOTTOMLEFT", 200*x, -10 - 30*y)
    checkBox.text:SetText(s)
    checkBox.key = k
    checkBox:SetScript("OnClick", function(self)
        local changes = optionsFrame.settingChanges.charaSettings
        changes.shownDebuffType = changes.shownDebuffType or {}
        changes.shownDebuffType[self.key] = self:GetChecked()
    end)

    optionsFrame.checkDebuffTypes[k] = checkBox
    counter = counter + 1
end

-- Broadcast settings
local prevText = CsAlert.func.createHeader(
    CsAlert.strings.broadcastSettings,
    optionsFrame.checkDebuffTypes[firstKey],
    0,
    -20 - 30*ceil(counter / 3),
    GameFontNormal
)

-- Available options for broadcast dropdowns
optionsFrame.instanceBroadcastTargets = {
    party = {
        "NONE",
        "SAY",
        "YELL",
        "PARTY"
    },
    raid = {
        "NONE",
        "SAY",
        "YELL",
        "PARTY",
        "RAID"
    },
    pvp = {
        "NONE",
        "SAY",
        "YELL",
        "PARTY",
        "INSTANCE_CHAT"
    }
}

-- Broadcast dropdowns
CsAlert.optionsFrame.dropDownBroadcast = {}
counter = 0
for instanceType, instanceString in pairs(CsAlert.strings.instanceTypes) do
    text = optionsFrame:CreateFontString(nil, "ARTWORK")
    text:SetFontObject(GameFontWhiteSmall)
    text:SetText(instanceString .. ":")
    text:SetPoint("TOPLEFT", prevText, "BOTTOMLEFT", 200*counter, -20)

    local dropDown = CreateFrame("Frame", "CsAlertDropDown"..instanceType, optionsFrame, "UIDropDownMenuTemplate")
    dropDown:SetPoint("TOPLEFT", text, "BOTTOMLEFT", -20, -5)
    dropDown.settingName = instanceType
    UIDropDownMenu_SetWidth(dropDown, 120)
    UIDropDownMenu_Initialize(dropDown, function(frame, level, menuList)
        local info = UIDropDownMenu_CreateInfo()
        info.func = function(self, arg1)
            UIDropDownMenu_SetText(frame, self.value)
            local changes = optionsFrame.settingChanges.charaSettings
            changes.broadcastSettings = changes.broadcastSettings or {}
            changes.broadcastSettings[frame.settingName] = arg1
        end
        info.minWidth = 120
        info.notCheckable = true
        info.checked = function(self)
            if optionsFrame.settingChanges.broadcastSettings == nil then
                return self.arg1 == CsAlert.charaSettings.broadcastSettings[frame.settingName]
            else
                return self.arg1 == optionsFrame.settingChanges.charaSettings.broadcastSettings[frame.settingName]
            end
        end
        for _, val in ipairs(optionsFrame.instanceBroadcastTargets[instanceType]) do
            info.text, info.arg1 = CsAlert.strings.broadcastTargets[val], val
            UIDropDownMenu_AddButton(info)
        end
    end)
    optionsFrame.dropDownBroadcast[instanceType] = dropDown
    counter = counter + 1
end

-- Author info
text = optionsFrame:CreateFontString(nil, "ARTWORK")
text:SetFontObject(GameFontDisableSmall)
text:SetText(CsAlert.strings.authorInfo)
text:SetPoint("BOTTOMRIGHT", optionsFrame, "BOTTOMRIGHT", -10, 10)


SpellDB.lua
local _, CsAlert = ...

local spellIds = {}

-- TODO: convert all these to spell IDs
-- Spells that interrupts and prevent spellcasting
spellIds.counterspell = {
    24687,  -- "Aspect of Jeklik",
    2139,   -- "Counterspell",
    10887,  -- "Crowd Pummel",
    8042,   -- "Earth Shock",
    19675,  -- "Feral Charge Effect",
    6730,   -- "Head Butt",
    1766,   -- "Kick",
    6552,   -- "Pummel",
    72,     -- "Shield Bash",
    19244,  -- "Spell Lock"
}

-- Spells that put the target to sleep
spellIds.asleep = {
    9256,   -- "Deep Sleep",
    8040,   -- "Druid's Slumber",
    16798,  -- "Enchanting Lullaby",
    7967,   -- "Naralex's Nightmare",
    700     -- "Sleep"
}

-- Spells that banish the target
spellIds.banish = {
    710,    -- "Banish",
    16451   -- "Judge's Gavel"
}

-- Spells that charms the target
spellIds.charm = {
    6358    -- "Seduction"
}

-- Spells that disorient the target
spellIds.disorient = {
    2094,   -- "Blind",
    26108,  -- "Glimpse of Madness",
    19503   -- "Scatter Shot"
}

-- Spells that cause the target to run in fear
spellIds.fear = {
    26641,  -- "Aura of Fear",
    18431,  -- "Bellowing Roar",
    21330,  -- "Corrupted Fear",
    6789,   -- "Death Coil",
    23275,  -- "Dreadful Fright",
    5782,   -- "Fear",
    25815,  -- "Frightening Shriek",
    5484,   -- "Howl of Terror",
    16508,  -- "Intimidating Roar",
    5246,   -- "Intimidating Shout",
    19577,  -- "Intimidation",
    19408,  -- "Panic",
    3109,   -- "Presence of Death",
    8122,   -- "Psychic Scream",
    21869,  -- "Repulsive Gaze",
    8225,   -- "Run Away!",
    7399,   -- "Terrify",
    8715,   -- "Terrifying Howl",
    14100,  -- "Terrifying Roar",
    6605,   -- "Terrifying Screech",
    21898,  -- "Warlock Terror",
    25260   -- "Wings of Despair"
}

-- Spells that incapacitate the target
spellIds.incapacitate = {
    1776,   -- "Gouge",
    22570,  -- "Mangle",
    13327,  -- "Reckless Charge",
    20066,  -- "Repentance",
    6770    -- "Sap"
}

-- Spells that mind controls the target
spellIds.mindcontrol = {
    12888,  -- "Cause Insanity",
    7645,   -- "Dominate Mind",
    13180,  -- "Gnomish Mind Control Cap",
    605,    -- "Mind Control",
    19469   -- "Poison Mind"
}

-- Spells that polymorph the target
spellIds.polymorph = {
    17738,  -- "Curse of the Plague Rat",
    22274,  -- "Greater Polymorph",
    11641,  -- "Hex",
    118,    -- "Polymorph",
    23603   -- "Wild Polymorph"
}

-- Spells that root the target
spellIds.root = {
    10852,  -- "Battle Net",
    113,    -- "Chains of Ice",
    5424,   -- "Claw Grasp",
    4246,   -- "Clenched Pinchers",
    19306,  -- "Counterattack",
    5219,   -- "Draw of Thistlenettle",
    8377,   -- "Earthgrab",
    11820,  -- "Electrified Net",
    4962,   -- "Encasing Webs",
    22994,  -- "Entangle",
    339,    -- "Entangling Roots",
    19185,  -- "Entrapment",
    15471,  -- "Enveloping Web",
    24110,  -- "Enveloping Webs",
    19675,  -- "Feral Charge Effect",
    10017,  -- "Frost Hold",
    122,    -- "Frost Nova",
    12494,  -- "Frostbite",
    8142,   -- "Grasping Vines",
    14030,  -- "Hooked Net",
    11264,  -- "Ice Blast",
    22519,  -- "Ice Nova",
    23694,  -- "Improved Hamstring",
    19229,  -- "Improved Wing Clip",
    5567,   -- "Miring Mud",
    8346,   -- "Mobility Malfunction",
    3542,   -- "Naraxis Web",
    6533,   -- "Net",
    13138,  -- "Net-o-Matic",
    23414,  -- "Paralyze",
    22935,  -- "Planted",
    22707,  -- "Root",
    7761,   -- "Shared Bonds",
    7295,   -- "Soul Drain",
    17333,  -- "Spider's Kiss",
    8312,   -- "Trap",
    745,    -- "Web",
    16469,  -- "Web Explosion",
    12252,  -- "Web Spray",
    24170,  -- "Whipweed Entangle"
    24152   -- "Whipweed Roots"
}

-- Spells that silence the target
spellIds.silence = {
    19821,  -- "Arcane Bomb",
    16838,  -- "Banshee Shriek",
    18469,  -- "Counterspell - Silenced",
    3589,   -- "Deafening Screech",
    18425,  -- "Kick - Silenced",
    6942,   -- "Overwhelming Stench",
    12946,  -- "Putrid Stench",
    7074,   -- "Screams of the Past",
    9552,   -- "Searing Flames",
    18498,  -- "Shield Bash - Silenced",
    15487,  -- "Silence",
    23918,  -- "Sonic Burst",
    19393,  -- "Soul Burn",
    19244   -- "Spell Lock"
}

-- Spells that stun the target
spellIds.stun = {
    24690,  -- "Aspect of Arlokk",
    24686,  -- "Aspect of Mar'li",
    6466,   -- "Axe Toss",
    6253,   -- "Backhand",
    5211,   -- "Bash",
    4067,   -- "Big Bronze Bomb",
    4069,   -- "Big Iron Bomb",
    15268,  -- "Blackout",
    17293,  -- "Burning Winds",
    7922,   -- "Charge Stun",
    6409,   -- "Cheap Shot",
    6945,   -- "Chest Pains",
    12809,  -- "Concussion Blow",
    16096,  -- "Cowering Roar",
    5403,   -- "Crash of Waves",
    17286,  -- "Crusader's Hammer",
    5106,   -- "Crystal Flash",
    3635,   -- "Crystal Gaze",
    3636,   -- "Crystalline Slumber",
    16104,  -- "Crystallize",
    19784,  -- "Dark Iron Bomb",
    18395,  -- "Dismounting Shot",
    21152,  -- "Earthshaker",
    25189,  -- "Enveloping Winds",
    7139,   -- "Fel Stomp",
    13902,  -- "Fist of Ragnaros",
    15743,  -- "Flamecrack",
    28323,  -- "Flameshocker's Revenge",
    28314,  -- "Flameshocker's Touch",
    16803,  -- "Flash Freeze",
    5276,   -- "Freeze",
    11836,  -- "Freeze Solid",
    17011,  -- "Freezing Claw",
    3355,   -- "Freezing Trap Effect",
    3143,   -- "Glacial Roar",
    13237,  -- "Goblin Mortar",
    12734,  -- "Ground Smash",
    19364,  -- "Ground Stomp",
    6524,   -- "Ground Tremor",
    6982,   -- "Gust of Wind",
    853,    -- "Hammer of Justice",
    19780,  -- "Hand of Ragnaros",
    6730,   -- "Head Butt",
    14102,  -- "Head Smash",
    12543,  -- "Hi-Explosive Bomb",
    20683,  -- "Highlord's Justice",
    11264,  -- "Ice Blast",
    16869,  -- "Ice Tomb",
    12355,  -- "Impact",
    20253,  -- "Intercept Stun",
    4068,   -- "Iron Grenade",
    408,    -- "Kidney Shot",
    20276,  -- "Knockdown",
    6266,   -- "Kodo Stomp",
    4065,   -- "Large Copper Bomb",
    25852,  -- "Lash",
    10856,  -- "Link Dead",
    13808,  -- "M73 Frag Grenade",
    5530,   -- "Mace Stun Effect",
    17500,  -- "Malown's Slam",
    12421,  -- "Mithril Frag Bomb",
    3609,   -- "Paralyzing Poison",
    11020,  -- "Petrify",
    9005,   -- "Pounce",
            -- "Psychic Scream",
            -- Need to figure out how to differentiate between the two effects
    18093,  -- "Pyroclasm",
    19641,  -- "Pyroclast Barrage",
    8285,   -- "Rampage",
    3446,   -- "Ravage",
    12798,  -- "Revenge Stun",
    6304,   -- "Rhahk'Zor Slam",
    4064,   -- "Rough Copper Bomb",
    17276,  -- "Scald",
    6927,   -- "Shadowstalker Slash",
    5918,   -- "Shadowstalker Stab",
    8242,   -- "Shield Slam",
    3551,   -- "Skull Crack",
    11430,  -- "Slam",
    4066,   -- "Small Bronze Bomb",
    6435,   -- "Smite Slam",
    8817,   -- "Smoke Bomb",
    24671,  -- "Snap Kick",
    16922,  -- "Starfire Stun",
            -- "Stomp",
            -- There's another spell named Stomp that causes a slow effect instead
    20685,  -- "Storm Bolt",
    19136,  -- "Stormbolt",
    56,     -- "Stun",
    16497,  -- "Stun Bomb",
    21188,  -- "Stun Bomb Attack",
    5648,   -- "Stunning Blast",
    15283,  -- "Stunning Blow",
    5703,   -- "Stunning Strike",
    5708,   -- "Swoop",
    23364,  -- "Tail Lash",
    12562,  -- "The Big One",
    19769,  -- "Thorium Grenade",
    21748,  -- "Thorn Volley",
    16075,  -- "Throw Axe",
    8150,   -- "Thundercrack",
    7803,   -- "Thundershock",
    835,    -- "Tidal Charm",
    21990,  -- "Tornado",
    3263,   -- "Touch of Ravenclaw",
    20549,  -- "War Stomp",
    24600,  -- "Web Spin",
    6749    -- "Wide Swipe"
}

-- Convert spell IDs to names (ensures the names match the locale)
CsAlert.spellTypeDb = {}
CsAlert.spellList = {}
for spellType, spells in pairs(spellIds) do
    CsAlert.spellTypeDb[spellType] = {}
    for _, spellId in ipairs(spells) do
        local spellName = GetSpellInfo(spellId)
        CsAlert.spellTypeDb[spellType][spellName] = true
        CsAlert.spellList[spellName] = true
    end
end


Main.lua
local _, CsAlert = ...

-- Default settings
CsAlert.globalSettings = {
    testMode = true,
    showAllDebuff = false
}
CsAlert.charaSettings = {
    minDebuffDuration = 3,
    shownDebuffType = {
        counterspell=true,
        asleep=true,
        banish=true,
        charm=true,
        disorient=true,
        fear=false,
        incapacitate=true,
        mindcontrol=true,
        polymorph=true,
        root=false,
        silence=true,
        stun=true
    },
    broadcastSettings = {
        party="SAY",
        raid="NONE",
        pvp="YELL"
    }
}
CsAlert_GlobalSettings = CsAlert_GlobalSettings or {}
CsAlert_CharaSettings = CsAlert_CharaSettings or {}

-- EVENT HANDLERS --
function CsAlert.handler(self, event, ...)
    CsAlert.onEvent[event](self, ...)
end

CsAlert.onEvent = {}

function CsAlert.onEvent.COMBAT_LOG_EVENT_UNFILTERED(self)
    local inInstance, instanceType = IsInInstance()

    -- In onEvent, SendChatMessage only works inside instances.
    if not (inInstance or CsAlert.testMode) then
        return
    end

    -- Check if current instance type is set to broadcast, and get broadcast target
    if CsAlert_CharaSettings.broadcastSettings[instanceType] == "NONE" and not CsAlert_GlobalSettings.showAllDebuff then
        return
    end
    local msgTarget = CsAlert_CharaSettings.broadcastSettings[instanceType]

    -- Get payload from CLEU
    local _, eventName, _, _, sourceName, _, sourceRaidFlags, destGUID = CombatLogGetCurrentEventInfo()
    local spellName = select(13, CombatLogGetCurrentEventInfo())

    -- Ignore all spells not targeted at the player
    if destGUID ~= UnitGUID("player") then
        return
    end

    -- Ignore spells not in the database
    if not CsAlert.spellList[spellName] and not CsAlert_GlobalSettings.showAllDebuff then
        return
    end

    -- Ignore spells whose category is disabled
    local ignoreSpell = true
    for spType, include in pairs(CsAlert_CharaSettings.shownDebuffType) do
        if include and CsAlert.spellTypeDb[spType][spellName] then
            ignoreSpell = false
            break
        end
    end
    if ignoreSpell then
        return
    end

    if eventName == "SPELL_INTERRUPT" then
        if CsAlert.spellTypeDb.counterspell[spellName] then
            local extraSpellName = select(16, CombatLogGetCurrentEventInfo())
            local raidFlag = CsAlert.func.getRaidTargetString(sourceRaidFlags)

            -- If GetSpellLossOfControlCooldown is queried now, it won't have the information available yet
            -- Using a different event trigger to get around this until a better way is found
            C_ChatInfo.SendAddonMessage("CsAlert", ("%s|%s|%s|%s|%s"):format(extraSpellName, spellName, sourceName, msgTarget, raidFlag), "SAY")
        end
    elseif eventName == "SPELL_AURA_APPLIED" or eventName == "SPELL_AURA_APPLIED_DOSE" or eventName == "SPELL_AURA_REFRESH" then
        -- Check what spell applied the aura
        -- We need to use UnitDebuff to get duration of aura though...
        local raidFlag = CsAlert.func.getRaidTargetString(sourceRaidFlags)
        local message

        for i = 1, 40 do
            local name, _, _, _, duration, _, source, _, _, spellId = UnitDebuff("player", 41 - i)
            -- `source` is the UnitId that casted the aura on the player...
            -- However, sometimes it returns nil, so it's not 100% reliable
            -- For now, we'll just check if the spell name matches.
            if name ~= nil and GetSpellInfo(spellId) == spellName then
                if not CsAlert_GlobalSettings.showAllDebuffs and duration < CsAlert_CharaSettings.minDebuffDuration then
                    return
                end

                if raidFlag ~= "" and raidFlag ~= nil then
                    message = ("Afflicted: %is %s from {%s} %s"):format(duration, name, raidFlag, sourceName)
                else
                    message = ("Afflicted: %is %s from %s"):format(duration, name, sourceName)
                end
                break
            end
        end

        if message then
            CsAlert.func.announce(message, msgTarget)
        end
    end
end

function CsAlert.onEvent.CHAT_MSG_ADDON(self, ...)
    local prefix, text, _, sender = ...

    -- Only accept messages from this addon, sent by player
    if prefix ~= "CsAlert" or sender ~= UnitName("player") .. '-' .. GetNormalizedRealmName() then
        return
    end

    -- Extract parameters back from the message
    local t = {}
    for s in text:gmatch("([^|]+)") do
        table.insert(t, s)
    end

    local extraSpellName, _, sourceName, msgTarget, raidFlag = unpack(t)
    local _, duration = GetSpellLossOfControlCooldown(extraSpellName)
    local message = ""
    if duration > 0 and duration >= CsAlert_CharaSettings.minDebuffDuration then
        if raidFlag ~= "" and raidFlag ~= nil then
            message = ("Afflicted: %is spell-lock from {%s} %s"):format(duration, raidFlag, sourceName)
        else
            message = ("Afflicted: %is spell-lock from %s"):format(duration, sourceName)
        end
    end

    CsAlert.func.announce(message, msgTarget)
end

function CsAlert.onEvent.ADDON_LOADED(self, addonName)
    if addonName == 'CounterspellAlert' then
        local globalSettings, charaSettings
        globalSettings = CsAlert.func.filterTable(CsAlert_GlobalSettings, CsAlert.globalSettings)
        charaSettings = CsAlert.func.filterTable(CsAlert_CharaSettings, CsAlert.charaSettings)
        CsAlert_GlobalSettings = CsAlert.func.deepCopy(CsAlert.globalSettings)
        CsAlert_CharaSettings = CsAlert.func.deepCopy(CsAlert.charaSettings)
        CsAlert.func.updateTable(CsAlert_GlobalSettings, globalSettings)
        CsAlert.func.updateTable(CsAlert_CharaSettings, charaSettings)

        CsAlert.func.updateOptions(CsAlert_GlobalSettings, CsAlert_CharaSettings)

        print("|cffff7f3f" .. CsAlert.strings.addonLoaded)
    end
end

-- Create main frame
local mainFrame = CreateFrame("Frame")
mainFrame:RegisterEvent("ADDON_LOADED")
mainFrame:RegisterEvent("CHAT_MSG_ADDON")
mainFrame:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
mainFrame:SetScript("OnEvent", CsAlert.handler)
C_ChatInfo.RegisterAddonMessagePrefix("CsAlert")

-- Assign slash command
SLASH_CSALERT1 = "/csalert"
function SlashCmdList.CSALERT(arg)
    local t = {}
    for s in arg:gmatch("([^ ]+)") do
        table.insert(t, s)
    end

    if t[1] == "dump" then
        if t[2] == "settings" then
            print("|cffff7f3f<CsAlert>|r " .. CsAlert.strings.dumpSettings)
            print(CsAlert.strings.globalSettings .. ":")
            CsAlert.func.dumpTable(CsAlert_GlobalSettings)
            print(CsAlert.strings.charaSettings .. ":")
            CsAlert.func.dumpTable(CsAlert_CharaSettings)
        elseif t[2] == "spells" then
            print("|cffff7f3f<CsAlert>|r " .. CsAlert.strings.dumpSpells)
            CsAlert.func.dumpTable(CsAlert.spellTypeDb)
        end
    else
        InterfaceOptionsFrame_OpenToCategory(CsAlert.optionsFrame)
    end
end
